File: apps/api/src/jobs/ImportFacebookInsights.js
import { findById as findAccount, updateAccount } from '../models/Account.js'
import { upsertFacebookInsight } from '../models/Metric.js'
import { getRedis } from '../db/redis.js'
import axios from 'axios'

const GRAPH = 'https://graph.facebook.com/v25.0'

// Facebook insight type constants (mirrors FacebookInsightType enum)
export const FacebookInsightType = {
  PAGE_POST_ENGAGEMENTS:  1,
  PAGE_POSTS_IMPRESSIONS: 2,
}

const METRIC_NAME_TO_TYPE = {
  page_post_engagements:  FacebookInsightType.PAGE_POST_ENGAGEMENTS,
  page_posts_impressions: FacebookInsightType.PAGE_POSTS_IMPRESSIONS,
}

export async function importFacebookInsightsJob({ account_id }) {
  const account = await findAccount(account_id)
  if (!account || !account.authorized || account.provider !== 'facebook') return

  const rateLimitKey = `agentmarket:ratelimit:facebook:${account_id}`
  const limited = await getRedis().get(rateLimitKey)
  if (limited) return

  const token = account.access_token?.token
  if (!token) return

  try {
    const since = new Date()
    since.setDate(since.getDate() - 90)

    // Mirrors getPageInsights(): metric=page_post_engagements,page_posts_impressions, period=day
    const res = await axios.get(`${GRAPH}/${account.provider_id}/insights`, {
      params: {
        access_token: token,
        metric: 'page_post_engagements,page_posts_impressions',
        period: 'day',
        since: since.toISOString().split('T')[0],
        until: new Date().toISOString().split('T')[0],
      },
    })

    const insights = res.data?.data || []

    for (const insight of insights) {
      const type = METRIC_NAME_TO_TYPE[insight.name]
      if (!type) continue

      for (const item of insight.values || []) {
        // end_time format: "2024-01-15T08:00:00+0000"
        const date = item.end_time?.split('T')[0] || new Date().toISOString().split('T')[0]
        await upsertFacebookInsight(account_id, type, item.value ?? 0, date)
      }
    }

    console.log(`[ImportFacebookInsights] account=${account_id} insight_types=${insights.length}`)
  } catch (err) {
    if (err.response?.status === 429) await getRedis().setex(rateLimitKey, 3600, '1')
    if (err.response?.status === 401 || err.response?.data?.error?.code === 190) {
      await updateAccount(account_id, { authorized: false })
      getRedis().publish('agentmarket:account_unauthorized', JSON.stringify({ account_id }))
    }
    throw err
  }
}

File: .ai-worker-analysis-dump/09-tasks/tasks/imports/import_facebook_insights.py
"""Facebook page insights — Graph via ``connectors.facebook``; persistence via apps/api."""

import logging
from datetime import datetime, timezone, timedelta

from bson import ObjectId
from bson.errors import InvalidId

from celery_app import celery_app
from connectors import facebook as facebook_connector
from db import get_redis
from lib import worker_api

logger = logging.getLogger(__name__)

RATE_KEY_PREFIX = 'agentmarket:ratelimit:facebook:'

METRIC_NAME_TO_TYPE = {
    'page_post_engagements': 1,
    'page_posts_impressions': 2,
}


@celery_app.task(name='tasks.imports.import_facebook_insights', bind=True, max_retries=3)
def import_facebook_insights(self, account_id: str):
    r = get_redis()
    try:
        ObjectId(account_id)
    except InvalidId:
        logger.warning('Invalid account_id: %s', account_id)
        return

    if not worker_api.configured():
        logger.warning('[import_facebook_insights] WORKER_API_SECRET not set — skipped')
        return

    try:
        account = worker_api.get_account_worker_context(account_id)
    except Exception:
        return

    if not account or not account.get('authorized') or account.get('provider') != 'facebook':
        return

    rate_key = f'{RATE_KEY_PREFIX}{account_id}'
    if r.get(rate_key):
        return

    token = (account.get('access_token') or {}).get('token')
    if not token:
        return

    pid = account.get('provider_id')
    if not pid:
        return

    since = (datetime.now(timezone.utc) - timedelta(days=90)).date().isoformat()
    until = datetime.now(timezone.utc).date().isoformat()

    result = facebook_connector.fetch_page_insights_day_series(
        token,
        str(pid),
        since=since,
        until=until,
    )

    if result.get('ok'):
        insights = result.get('insights') or []
        for insight in insights:
            name = insight.get('name')
            itype = METRIC_NAME_TO_TYPE.get(name)
            if not itype:
                continue
            for item in insight.get('values') or []:
                end_time = item.get('end_time') or ''
                date = end_time.split('T')[0] if end_time else until
                val = item.get('value')
                if val is None:
                    val = 0
                worker_api.upsert_facebook_insight(account_id, itype, val, date)

        logger.info('[import_facebook_insights] account=%s insight_types=%s', account_id, len(insights))
        return

    if result.get('rate_limited'):
        r.setex(rate_key, 3600, '1')
        logger.warning('[import_facebook_insights] 429 %s', account_id)
        return
    if result.get('unauthorized'):
        try:
            worker_api.patch_account_deauthorized(account_id)
        except Exception as exc:
            logger.error('[import_facebook_insights] deauthorize failed: %s', exc)
        logger.warning('[import_facebook_insights] unauthorized %s', account_id)
        return

    sc = int(result.get('status_code') or 0)
    logger.error('[import_facebook_insights] account=%s status=%s', account_id, sc)
    raise self.retry(exc=RuntimeError(result.get('error', 'Graph error')), countdown=300) from None
